subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ResolveInputChangesStep.java

```diff
@@ -0,0 +1,91 @@
+
+import org.gradle.internal.execution.IncrementalChangesContext;
+import org.gradle.internal.execution.InputChangesContext;
+import org.gradle.internal.execution.Result;
+import org.gradle.internal.execution.Step;
+import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.execution.history.AfterPreviousExecutionState;
+import org.gradle.internal.execution.history.BeforeExecutionState;
+import org.gradle.internal.execution.history.changes.InputChangesInternal;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Optional;
+
+public class ResolveInputChangesStep<C extends IncrementalChangesContext> implements Step<C, Result> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResolveInputChangesStep.class);
+
+    private final Step<? super InputChangesContext, ? extends Result> delegate;
+
+    public ResolveInputChangesStep(Step<? super InputChangesContext, ? extends Result> delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Result execute(C context) {
+        final UnitOfWork work = context.getWork();
+        Optional<InputChangesInternal> inputChanges = work.isRequiresInputChanges()
+            ? Optional.of(determineInputChanges(work, context))
+            : Optional.empty();
+        return delegate.execute(new InputChangesContext() {
+            @Override
+            public Optional<String> getRebuildReason() {
+                return context.getRebuildReason();
+            }
+
+            @Override
+            public Optional<AfterPreviousExecutionState> getAfterPreviousExecutionState() {
+                return context.getAfterPreviousExecutionState();
+            }
+
+            @Override
+            public Optional<BeforeExecutionState> getBeforeExecutionState() {
+                return context.getBeforeExecutionState();
+            }
+
+            @Override
+            public Optional<InputChangesInternal> getInputChanges() {
+                return inputChanges;
+            }
+
+            @Override
+            public boolean isIncrementalExecution() {
+                return inputChanges.map(changes -> changes.isIncremental()).orElse(false);
+            }
+
+            @Override
+            public UnitOfWork getWork() {
+                return work;
+            }
+        });
+    }
+
+    private InputChangesInternal determineInputChanges(UnitOfWork work, IncrementalChangesContext context) {
+        return context.getChanges()
+            .map(changes -> {
+                InputChangesInternal inputChanges = changes.createInputChanges();
+                if (!inputChanges.isIncremental()) {
+                    LOGGER.info("All input files are considered out-of-date for incremental {}.", work.getDisplayName());
+                }
+                return inputChanges;
+            })
+            .orElseThrow(() -> new UnsupportedOperationException("Cannot use input changes when input tracking is disabled."));
```

Comment:
I think in this case `get()` is okay. A more specific exception when the unthinkable happens doesn't add much, yet it makes the code more complex.

Reply:
Changed to `get()`.